tools/generate_release_notes.py

```diff
@@ -1,246 +1,331 @@
-"""Generate the release notes automatically from Github pull requests.
-
-Start with:
-```
-export GH_TOKEN=<your-gh-api-token>
-```
-
-Then, for a major release:
-```
-python /path/to/generate_release_notes.py v0.14.0 main --version 0.15.0
-```
-
-For a minor release:
-```
-python /path/to/generate_release_notes.py v.14.2 v0.14.x --version 0.14.3
-```
-
-You should probably redirect the output with:
-```
-python /path/to/generate_release_notes.py [args] | tee release_notes.rst
-```
-
-You'll require PyGitHub and tqdm, which you can install with:
-```
-pip install -r requirements/_release_tools.txt
-```
-
-References
-https://github.com/scikit-image/scikit-image/issues/3404
-https://github.com/scikit-image/scikit-image/issues/3405
-"""
+"""Generate release notes automatically from GitHub pull requests."""
+
+
 import os
+import re
+import sys
 import argparse
-from datetime import datetime
-from warnings import warn
-
-from github import Github
-
-try:
-    from tqdm import tqdm
-except ImportError:
-    warn(
-        "tqdm not installed. This script takes approximately 5 minutes "
-        "to run. To view live progressbars, please install tqdm. "
-        "Otherwise, be patient."
-    )
+import logging
+from pathlib import Path
+from dataclasses import dataclass
+from typing import Callable
+from collections.abc import Iterable
+
+import requests_cache
+from tqdm import tqdm
+from github import Github, Repository, PullRequest, NamedUser, Commit
 
-    def tqdm(i, **kwargs):
-        return i
 
+logger = logging.getLogger(__name__)
 
-GH_USER = "scikit-image"
+here = Path(__file__).parent
+
+GH_URL = "https://github.com"
+GH_ORG = "scikit-image"
 GH_REPO = "scikit-image"
-GH_TOKEN = os.environ.get("GH_TOKEN")
-if GH_TOKEN is None:
-    raise RuntimeError(
-        "It is necessary that the environment variable `GH_TOKEN` "
-        "be set to avoid running into problems with rate limiting. "
-        "One can be acquired at https://github.com/settings/tokens.\n\n"
-        "You do not need to select any permission boxes while generating "
-        "the token."
-    )
 
-g = Github(GH_TOKEN)
-repository = g.get_repo(f"{GH_USER}/{GH_REPO}")
 
+def lazy_tqdm(*args, **kwargs):
+    """Defer initialization of progress bar until first item is requested."""
+    kwargs["file"] = kwargs.get("file", sys.stderr)
+    yield from tqdm(*args, **kwargs)
+
+
+def commits_between(repo: Repository, start_rev: str, stop_rev: str) -> set[Commit]:
+    """Fetch commits between two revisions excluding the commit of `start_rev`."""
+    # https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits
+    comparison = repo.compare(base=start_rev, head=stop_rev)
+    return set(comparison.commits)
+
+
+def pull_requests_from_commits(commits: Iterable[Commit]) -> set[Commit]:
```

Comment:
`pull_request_from_commits` returns `set[Commit]`? A bit counter-intuitive, especially given the docstring.

Reply:
Yeah, very counter-intuitive because its a bug / typo! :sweat_smile: Good catch.